### ISSUE TITLE
Handle requirements on different `node` versions

### ISSUE THREAD

**Issue by dgutson** [oss-review-toolkit/ort#6694]
We get this error:
```
EBADENGINE Unsupported engine {
EBADENGINE   package: 'xxx@2.16.0',
EBADENGINE   required: { node: '>=16 <18' },
EBADENGINE   current: { node: 'v18.12.1', npm: '8.15.1' }
EBADENGINE }
```

Does ORT use the _installed_ node version? What if it isn't installed? I think it should not use whatever is installed, but the required version. I think ORT should use https://github.com/tj/n which is what node developers use to use different node versions.

This is specially important for multi-repo systems using different versions of node.

**Comment by sschuberth** [comment]
> Does ORT use the _installed_ node version?

Yes, ORT uses the `node` version in PATH (the same is true for all external tools ORT depends on). But depending on how you run ORT that may mean different things. If you run ORT from the Docker image, the `node` version (currently [18.14.2](https://github.com/oss-review-toolkit/ort/blob/2e78d0f4f756b39afb61c8f1d542e5080849ef90/Dockerfile#L205)) from the Docker image will be used, not the one from the host.

> What if it isn't installed?

Analysis will fail completely. ORT will not try to bootstrap `node` or any other tool.

> I think ORT should use https://github.com/tj/n which is what node developers use to use different node versions.

The combination of [bootstrapping `n`](https://github.com/tj/n#installation) plus [running `n auto` or `n engine`](https://github.com/tj/n#specifying-nodejs-versions) indeed looks interesting.

> This is specially important for multi-repo systems using different versions of node.

Which means the bootstrapping would need to happen in `beforeResolution`, not once in the Docker image.

**Comment by nnobelis** [comment]
I would like to bring some news back in this ticket and I am willing to do the implementation.

My understanding is the following. We need to make `node` version configurable.

- The user defines the desired node version in their `.ort.yml`, in the "npm" package manager plugins options
- ORT installs the desired node version with either `n` or `nvm` (to determine, `nvm` is already present in ORT's Docker file)
- ORT calls a command to make the installed node version the default on the PATH
- NPM package manager run with the installed node version

Some remarks:
- Corepack is not related to this change, as is it used for Yarn and PNPM installation
- NPM version cannot be configured separately, but it can be controlled by choosing the right node version, since NPM is bundled with node (https://nodejs.org/en/about/previous-releases, one can see the NPM version by clicking on "details").
- To clarify: what is the impact of making node configurable on PNPM, Yarn and Yarn2 package managers ? Do they also need a configurable node version ?

**Comment by sschuberth** [comment]
> * ORT installs the desired node version with either `n` or `nvm` (to determine, `nvm` is already present in ORT's Docker file)

As mentioned by @mnonnenmacher, https://github.com/Schniz/fnm could also be an option.
